components/cover/tormatic.rst

```diff
@@ -0,0 +1,101 @@
+Tormatic/Novoferm Cover
+=======================
+
+.. seo::
+    :description: Setting up Tormatic and Novoferm covers in ESPHome.
+    :image: tormatic.png
+
+The ``tormatic`` cover platform allows you to control Tormatic and Novoferm
+garage door drives manufactured in 2016 onwards. The following models should
+be supported:
+
+ * Tormatic Black 600, Black 800, Black 1000
+ * Novoferm Novomatic 423
+ * Novoferm Novomatic 563 S
+ * Novoferm Novoport IV
+
+.. figure:: images/tormatic.png
+    :align: center
+
+    Tormatic Black 600 Garage Door Drive
+
+.. note::
+
+    This component was written and tested against a Novoferm 423, so
+    compatibility with other models is currently not guaranteed. Please update
+    this documentation if you confirm this to work on other models as well.
+    
+    The vendor's cloud-based remote is not known to perform model or feature
+    detection, so it's assumed that the protocol is the same across all models.
+
+.. code-block:: yaml
+
+    # Example configuration entry
+    uart:
+      rx_pin: GPIO19
+      tx_pin: GPIO20
+      baud_rate: 9600
+
+    cover:
+      - platform: tormatic
+        device_class: garage
+        name: Novoferm 423
+
+Setup
+-----
+
+The garage door drive ('unit') contains a female USB Type B port that does **not**
+speak the USB protocol. It is not a conventional USB port, though it supplies 5V
+between the usual USB Vcc and ground pins and can be used to feed an ESP32 board.
+
+The D+ and D- terminals are used as a 9600/8N1 UART pair **with 5V logic levels
+(!)**. All current ESP family devices expect 3V logic levels, so a level shifter
+is needed on the data lines for this to work correctly. Experiments with
+connecting an ESP32 S2's GPIO hardware directly to these pins yielded strange
+results; the ESP runs hot, doesn't connect to WiFi and the ESP's 3V tx logic
+level is slightly underpowered, so the unit doesn't receive the ESP's messages
+correctly most of the time.
+
+The unit pulls the USB's D- (white) line high and uses it to transmit data. It
+receives data on the D+ (green) line, meaning **the ESP should transmit on D+
+and receive on D-**. On a Wemos S2, these lines are directly routed to GPIO19
+(D-) and GPIO20 (D+), since there is no separate in-line USB chip. This makes
+the S2 an ideal device for this purpose; it keeps the cable simple and compact,
+since it needs to fit in a tight space in the unit. Only a single header pin
+needs to be soldered the PCB to supply 3.3V to the logic level shifter, but it
+can be bent 90 degrees to sit parallel to the PCB, keeping a low profile.
```

Comment:
This seems like the perfect opportunity to create a custom PCB with a male USB-B port that just attaches directly and has an ESP onboard.

Reply:
It sure is! Different kind of effort required, though. :) It would be tricky to fit that into the unit, the manufacturer's cloud-based dongle has level shifters on board and has a USB-B cable. The dongle goes into a separate compartment.